libs/runtime/ebpf_ring_buffer.c

```diff
@@ -135,78 +70,68 @@ _Must_inspect_result_ ebpf_result_t
     }
 
-    record->header.discarded = 0;
-    record->header.locked = 0;
-    memcpy(record->data, data, length);
-    result = EBPF_SUCCESS;
-Done:
-    ebpf_lock_unlock(&ring->lock, state);
-    return result;
+    memcpy(buffer, data, length);
+
+    return ebpf_ring_buffer_submit(buffer);
 }
 
 void
 ebpf_ring_buffer_query(_In_ ebpf_ring_buffer_t* ring, _Out_ size_t* consumer, _Out_ size_t* producer)
 {
-    ebpf_lock_state_t state = ebpf_lock_lock(&ring->lock);
-    *consumer = ring->consumer_offset;
-    *producer = ring->producer_offset;
-    ebpf_lock_unlock(&ring->lock, state);
+    *consumer = (size_t)ReadAcquire64(&ring->consumer_offset);
+    *producer = (size_t)ReadAcquire64(&ring->producer_offset);
 }
 
 _Must_inspect_result_ ebpf_result_t
 ebpf_ring_buffer_return(_Inout_ ebpf_ring_buffer_t* ring, size_t length)
 {
     EBPF_LOG_ENTRY();
-    ebpf_result_t result;
-    ebpf_lock_state_t state = ebpf_lock_lock(&ring->lock);
-    size_t local_length = length;
-    size_t offset = _ring_get_consumer_offset(ring);
-
-    if ((length > _ring_get_length(ring)) || length > _ring_get_used_capacity(ring)) {
-        EBPF_LOG_MESSAGE_UINT64_UINT64(
-            EBPF_TRACELOG_LEVEL_ERROR,
-            EBPF_TRACELOG_KEYWORD_MAP,
-            "ebpf_ring_buffer_return: Buffer too large",
-            ring->producer_offset,
-            ring->consumer_offset);
-        result = EBPF_INVALID_ARGUMENT;
-        goto Done;
-    }
+    int64_t length_remaining = (int64_t)length;
+
+    for (;;) {
+        int64_t producer_offset = ReadAcquire64(&ring->producer_offset);
+        int64_t consumer_offset = ReadAcquire64(&ring->consumer_offset);
 
-    // Verify count.
-    while (local_length != 0) {
-        ebpf_ring_buffer_record_t* record = _ring_record_at_offset(ring, offset);
-        if (local_length < record->header.length) {
+        if (length_remaining > (producer_offset - consumer_offset)) {
+            return EBPF_INVALID_ARGUMENT;
+        }
+
+        if (length_remaining == 0) {
             break;
         }
-        offset += record->header.length;
-        local_length -= record->header.length;
-    }
-    // Did it end on a record boundary?
-    if (local_length != 0) {
-        EBPF_LOG_MESSAGE_UINT64(
-            EBPF_TRACELOG_LEVEL_ERROR,
-            EBPF_TRACELOG_KEYWORD_MAP,
-            "ebpf_ring_buffer_return: Invalid buffer length",
-            local_length);
-        result = EBPF_INVALID_ARGUMENT;
-        goto Done;
-    }
 
-    _ring_advance_consumer_offset(ring, length);
-    result = EBPF_SUCCESS;
+        ebpf_ring_buffer_record_t* record =
+            (ebpf_ring_buffer_record_t*)(ring->shared_buffer + (consumer_offset & ring->mask));
 
-Done:
-    ebpf_lock_unlock(&ring->lock, state);
-    EBPF_RETURN_RESULT(result);
+        long size = ReadAcquire(&record->size);
+
+        // Can't return a locked record.
+        if (size & EBPF_RING_BUFFER_RECORD_FLAG_LOCKED) {
+            continue;
```

Comment:
Is there a reason not to just spin on the lock bit of the current record (rather than jumping to the top of the loop)?

Also, since there is only a single consumer (producer_offset-consumer_offset) should only increase, so I think the length_remaining check could be before the loop.

Reply:
Yes, there is a very good reason not to spin on the current bit. The bit will remain set until the submit is called, but as soon as the reserve completes there is a new next record to spin on. Hence looping around and checking the new producer offset.

Reply:
This is in ring_buffer_return, so that will remain the next record to return until we return it.